### PR TITLE
Crosswords: sticky clues aka dream clues

### PR DESCRIPTION
**Pull request by OliverJAsh**
top-to-bottom

iOS doesn't support sticky things when the keyboard is open, so we use absolute positioning and programatically update the value of top. Thus, it's not quite as smooth for those devices. Simulated here:

ios

### CODE REVIEW

File: static/src/javascripts/es6/projects/common/modules/crosswords/hidden-input.js [hunk 1]
```diff
@@ -20,7 +21,7 @@ export default class HiddenInput extends React.Component {
         if (detect.isBreakpoint({ max: 'mobile' })) {
             fastdom.read(function () {
                 const offsets = bonzo(React.findDOMNode(this.refs.input)).offset();
-                scroller.scrollTo(offsets.top - (offsets.height * 1.5), 250, 'easeOutQuad');
+                scroller.scrollTo(offsets.top - (offsets.height * 1.5) - $('.crossword__sticky-clue').offset().height, 250, 'easeOutQuad');
```
Comment (OliverJAsh) [inline static/src/javascripts/es6/projects/common/modules/crosswords/hidden-input.js]: :unamused: I hate that the sticky clue leaks into the input field component. Any better ideas?

**Comment by ScottPainterGNM** [comment]
Olly's stupendous dream clue™

**Comment by NathanielBennett** [comment]
:+1:

**Comment by OliverJAsh** [comment]
If you want to test what happens to fixed elements when the keyboard opens in iOS: http://output.jsbin.com/sejecu

File: static/src/javascripts/es6/projects/common/modules/crosswords/hidden-input.js [hunk 2]
```diff
@@ -18,10 +19,10 @@ export default class HiddenInput extends React.Component {
 
     componentDidUpdate () {
         if (detect.isBreakpoint({ max: 'mobile' })) {
-            fastdom.read(function () {
+            fastdom.read(() => {
                 const offsets = bonzo(React.findDOMNode(this.refs.input)).offset();
-                scroller.scrollTo(offsets.top - (offsets.height * 1.5), 250, 'easeOutQuad');
-            }.bind(this));
+                scroller.scrollTo(offsets.top - (offsets.height * 1.5) - $('.crossword__sticky-clue').offset().height, 250, 'easeOutQuad');
```
Comment (OliverJAsh) [inline static/src/javascripts/es6/projects/common/modules/crosswords/hidden-input.js]: I hate that this component has to core about the sticky clue. Any better ideas?

File: static/src/javascripts/es6/projects/common/modules/crosswords/main.js [hunk 3]
```diff
@@ -505,30 +546,50 @@ class Crossword extends React.Component {
             <AnagramHelper clue={focussed} grid={this.state.grid} close={this.onToggleAnagramHelper}/>
         );
 
+        const focussedClue = this.clueInFocus();
```
Comment (desbo) [inline static/src/javascripts/es6/projects/common/modules/crosswords/main.js]: we already have this a few lines up, as `focussed`

File: static/src/javascripts/es6/projects/common/modules/crosswords/main.js [hunk 4]
```diff
@@ -72,6 +75,44 @@ class Crossword extends React.Component {
...
+
+                    if (scrollY > (crosswordOffsetBottom - $stickyClueWrapper.offset().height)) {
+                        $stickyClueWrapper
+                            .css('top', 'auto')
+                            .css('bottom', 0);
+                    } else {
+                        // iOS doesn't support sticky things when the keyboard
+                        // is open, so we use absolute positioning and
+                        // programatically update the value of top
+                        if (detect.isIOS()) {
```
Comment (desbo) [inline static/src/javascripts/es6/projects/common/modules/crosswords/main.js]: don't think we need to check for this on every scroll event, i'd check for it outside of the hander and bind either a ios or non-ios function to scroll there

**Comment by OliverJAsh** [comment]
I've had a go at improving the performance in commit https://github.com/guardian/frontend/commit/89274beec8984e687e2eaa779d41ae5117751d17 and PR https://github.com/guardian/frontend/pull/10430.

Here is the difference:


It's still a big laggy, but I'm not really sure why that is. I'm not sure we can make it 100% smooth. Here is a simple playground version where I've been experimenting: http://output.jsbin.com/banaho/14/quiet

I'm not going to block this on the scrolling performance, but very keen to understand further, so please shout if you have ideas!

**Pull request merged**